### PR TITLE
 Add support for documenting all members of the workspace with "doc --all"

### PR DESCRIPTION
**Pull request by sdroege**
https://github.com/rust-lang/cargo/issues/3491

### CODE REVIEW

File: tests/doc.rs [hunk 1]
```diff
@@ -612,3 +613,97 @@ fn plugins_no_use_target() {
...
+            pub fn bar() {}
+        "#);
+    p.build();
+
+    // The order in which bar is compiled or documented is not deterministic
+    assert_that(p.cargo_process("doc")
+                 .arg("--all"),
+                execs().with_stderr_contains("[..] Documenting bar v0.1.0 ([..])")
+                       .with_stderr_contains("[..] Compiling bar v0.1.0 ([..])")
+                       .with_stderr_contains("[..] Documenting foo v0.1.0 ([..])"));
```
Comment (alexcrichton) [inline tests/doc.rs]: Could you add `.with_status(0)` assertions to these as well to ensure we test that the process ran successfully?

File: src/cargo/ops/cargo_doc.rs [hunk 2]
```diff
@@ -13,45 +13,53 @@ pub struct DocOptions<'a> {
...
 
     let mut lib_names = HashSet::new();
     let mut bin_names = HashSet::new();
-    if spec.is_empty() {
-        for target in package.targets().iter().filter(|t| t.documented()) {
-            if target.is_lib() {
-                assert!(lib_names.insert(target.crate_name()));
-            } else {
-                assert!(bin_names.insert(target.crate_name()));
+    if specs.is_empty() {
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_doc.rs]: I think this may actually work best if we remove this condition. I don't quite recall why it exists, but I think it's best if we execute the below logic for all packages, not just the root package.

File: tests/doc.rs (hunk 1 above)
Comment (sdroege) [inline tests/doc.rs]: Sure, but shouldn't this also be added to the "build --all" test? I'll send another pull request for the change of the other test then

Comment (sdroege) [inline tests/doc.rs]: I guess it would be "nice" but not necessary, as for building the test checks for the "finished successfully" line in the output

File: src/cargo/ops/cargo_doc.rs (hunk 2 above)
Comment (sdroege) [inline src/cargo/ops/cargo_doc.rs]: I agree, I only kept it because I thought someone probably had a reason for having done that and I just don't understand the reason :)

**Comment by sdroege** [comment]
Thanks for the review! Updated for both your comments

File: tests/doc.rs (hunk 1 above)
Comment (alexcrichton) [inline tests/doc.rs]: Ah yeah it's true I should have caught it on that PR! Want to send another PR for that?

Comment (sdroege) [inline tests/doc.rs]: Sure, will send something for that tomorrow :)
